### PR TITLE
[improve][broker] PIP-192 Moved the common broker load data feature(weightedMaxEMA) to BrokerLoadData

### PR DESCRIPTION
**Pull request by heesung-sn**
Master Issue: https://github.com/apache/pulsar/issues/16691

### Motivation

We will start raising PRs to implement PIP-192, https://github.com/apache/pulsar/issues/16691

### Modifications
This PR moved the common broker load data feature(`weightedMaxEMA`) to `BrokerLoadData`.

As `weightedMaxEMA` is available in the `BrokerLoadData` by this change, this PR
- removed `brokerAvgResourceUsageWithWeight` in LeastResourceUsageWithWeight -- we don't need to separately compute and store `weightedMaxEMA` in `LeastResourceUsageWithWeight and TransferShedder`. 

- refactored BrokerLoadData to precompute the load data features.

- updated the unit tests.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

  -  Updated unit tests.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

We will have separate PRs to update the Doc later.

### Matching PR in forked repository

PR in forked repository: https://github.com/heesung-sn/pulsar/pull/20

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/strategy/LeastResourceUsageWithWeight.java [hunk 1]
```diff
@@ -43,21 +41,17 @@ public class LeastResourceUsageWithWeight implements BrokerSelectionStrategy {
...
     }
 
     // A broker's max resource usage with weight using its historical load and short-term load data with weight.
     private double getMaxResourceUsageWithWeight(final String broker, final BrokerLoadData brokerLoadData,
                                                  final ServiceConfiguration conf, boolean debugMode) {
         final double overloadThreshold = conf.getLoadBalancerBrokerOverloadedThresholdPercentage() / 100.0;
-        final double maxUsageWithWeight =
-                updateAndGetMaxResourceUsageWithWeight(broker, brokerLoadData, conf, debugMode);
-
+        final var maxUsageWithWeight = brokerLoadData.getWeightedMaxEMA();
```
Comment (Demogorgon314) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/strategy/LeastResourceUsageWithWeight.java:54]: The `weightedMaxEMA` can be null. Should we handle the null case here?

**Review by Demogorgon314 — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/data/BrokerLoadData.java [hunk 2]
```diff
@@ -95,17 +124,35 @@ private void updateSystemResourceUsage(final ResourceUsage cpu, final ResourceUs
...
+    private double getMaxResourceUsageWithWeight(final double cpuWeight, final double memoryWeight,
                                                 final double directMemoryWeight, final double bandwidthInWeight,
                                                 final double bandwidthOutWeight) {
         return LocalBrokerData.max(cpu.percentUsage() * cpuWeight, memory.percentUsage() * memoryWeight,
                 directMemory.percentUsage() * directMemoryWeight, bandwidthIn.percentUsage() * bandwidthInWeight,
                 bandwidthOut.percentUsage() * bandwidthOutWeight) / 100;
     }
 
+
+    private void updateWeightedMaxEMA(ServiceConfiguration conf) {
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/data/BrokerLoadData.java:147]: What does EMA mean? Is it better to add a description for "EMA"?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/strategy/LeastResourceUsageWithWeight.java [hunk 3]
```diff
@@ -80,40 +74,6 @@ private double getMaxResourceUsageWithWeight(final String broker, final BrokerLo
...
-        if (debugMode) {
-            log.info(
-                    "Broker {} get max resource usage with weight: {}, history resource percentage: {}%, CPU weight: "
-                            + "{}, MEMORY weight: {}, DIRECT MEMORY weight: {}, BANDWIDTH IN weight: {}, BANDWIDTH "
-                            + "OUT weight: {} ",
-                    broker, historyUsage, historyPercentage, conf.getLoadBalancerCPUResourceWeight(),
-                    conf.getLoadBalancerMemoryResourceWeight(), conf.getLoadBalancerDirectMemoryResourceWeight(),
-                    conf.getLoadBalancerBandwithInResourceWeight(),
-                    conf.getLoadBalancerBandwithOutResourceWeight());
-        }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/strategy/LeastResourceUsageWithWeight.java:113]: Should the debug log be removed?

Comment (heesung-sn) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/strategy/LeastResourceUsageWithWeight.java:113]: No. I think we can keep this.

Added BrokerLoadData.toString() and added the debug log.

```
else if (debugMode) {
            log.info("Broker {} load data:{{}}", broker, brokerLoadData.toString(conf));
        }

```

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**